### PR TITLE
[To new_sync][IOTDB-2272] implement customized sync process: collector

### PR DESCRIPTION
**Pull request by Cpaulyz**
This pr contains: 
* Collector framework and IT.
* Refactor `PipeData` serialize and deserialze and add UT.
* Add `SyncConstant` and `SyncPathUtil `

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java [hunk 1]
```diff
@@ -19,4 +19,155 @@
...
+  public void startPipe(String pipeName, String remoteIp, long createTime) {
+    task.addScanDir(SyncPathUtil.getReceiverPipeLogDir(pipeName, remoteIp, createTime));
+  }
+
+  public void stopPipe(String pipeName, String remoteIp, long createTime) {
+    task.removeScanDir(SyncPathUtil.getReceiverPipeLogDir(pipeName, remoteIp, createTime));
+  }
+
+  private class ScanTask implements Runnable {
+    private final Set<String> scanPathSet;
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java:75]: concurrent error risk

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java [hunk 2]
```diff
@@ -19,4 +19,155 @@
...
+                  }
+                } catch (Exception e) {
+                  // TODO: how to response error message to sender?
+                  // TODO: should drop this pipe?
+                  logger.error(
+                      "Cannot load pipeData with serialize number {} and type {}, because {}",
+                      pipeData.getSerialNumber(),
+                      pipeData.getType(),
+                      e.getMessage());
+                  break;
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java:156]: somtimes we can continue

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java [hunk 3]
```diff
@@ -19,4 +19,155 @@
...
+                  }
+                  raf.close();
+                }
+                outputStream = new DataOutputStream(new FileOutputStream(recordFile, true));
+              } else {
+                // TODO: get buffer from transport
+                // pipeDataList = transport.getBufferPipeDataList();
+              }
+              boolean allLoaded = true;
+              for (int i = startNumber + 1; i < pipeDataList.size(); i++) {
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java:133]: pipeDataList from buffer may changes by time.

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/ReceiverService.java [hunk 4]
```diff
@@ -74,23 +78,39 @@ public boolean stopPipeServer() {
...
+  public void stopPipe(String pipeName, String remoteIp, long createTime) throws IOException {
     receiverManager.stopPipe(pipeName, remoteIp);
+    collector.stopPipe(pipeName, remoteIp, createTime);
   }
 
   /** drop an existed pipe named pipeName */
-  public void dropPipe(String pipeName, String remoteIp) throws IOException {
+  public void dropPipe(String pipeName, String remoteIp, long createTime) throws IOException {
     receiverManager.dropPipe(pipeName, remoteIp);
+    collector.stopPipe(pipeName, remoteIp, createTime);
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/ReceiverService.java:102]: should delete all files

File: server/src/main/java/org/apache/iotdb/db/newsync/pipedata/PipeData.java [hunk 5]
```diff
@@ -53,39 +58,30 @@ public long getSerialNumber() {
...
       case DELETION:
-        pipeData = new DeletionPipeData(serialNumber);
-        break;
+        return DeletionPipeData.deserialize(stream);
       case PHYSICALPLAN:
-        pipeData = new SchemaPipeData(serialNumber);
-        break;
+        return SchemaPipeData.deserialize(stream);
+      default:
+        logger.error("Deserialize PipeData error because Unknown type {}.", type);
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/pipedata/PipeData.java:79]: throw a UnsupportException may be better

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/recovery/ReceiverLog.java [hunk 6]
```diff
@@ -31,7 +31,7 @@
   private BufferedWriter bw;
 
   public ReceiverLog() throws IOException {
-    File logFile = new File(SenderConf.sysDir, SyncConstant.RECEIVER_LOG_NAME);
+    File logFile = new File(SyncPathUtil.getSysDir(), SyncConstant.RECEIVER_LOG_NAME);
     if (!logFile.getParentFile().exists()) {
       logFile.getParentFile().mkdirs();
     }
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/recovery/ReceiverLog.java:37]: mkdirs too early

Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/receiver/recovery/ReceiverLog.java:37]: fixed

File: server/src/main/java/org/apache/iotdb/db/newsync/pipedata/PipeData.java (hunk 5 above)
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/pipedata/PipeData.java:79]: fixed

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java (hunk 3 above)
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java]: refactor thread logic

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java (hunk 1 above)
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java]: refactor thread logic

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java (hunk 2 above)
Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java]: refactor thread logic

File: server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java [hunk 7]
```diff
@@ -19,4 +19,189 @@
...
+/** scan sync receiver folder and load pipeData into IoTDB */
+public class Collector {
+
+  private static final Logger logger = LoggerFactory.getLogger(Collector.class);
+  private static final int WAIT_TIMEOUT = 2000;
+  private ExecutorService executorService;
+  private Map<String, Future> taskFutures;
+
+  public Collector() {
+    taskFutures = new HashMap<>();
```
Comment (yschengzi) [inline server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java:50]: HashMap is not thread-safe. use concurrentHashMap plz.

Comment (Cpaulyz) [reply server/src/main/java/org/apache/iotdb/db/newsync/receiver/collector/Collector.java]: fixed

**Review by yschengzi — APPROVED** [review]
LGTM

**Pull request merged**